<a href="https://colab.research.google.com/github/joiepark/assignment_esaa/blob/main/YB%202.1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 데이터 핸들링 - 판다스



*   판다스는 행과 열로 이뤄진 2차원 데이터를 효울적으로 가공하고 처리할 수 있는 다양한 기능을 제공한다. 일반적으로 대부분의 데이터 세트는 2차원 데이터로, 행(Row)과 열(Column)로 구성되어 있다. 행과 열의 2차원 데이터는 인간이 가장 이해하기 쉬운 데이터 구조이면서도 효과적으로 데이터를 담을 수 있는 구조이기에 오늘날 가장 인기가 많다.
*   저수준 API가 대부분인 넘파이의 데이터 핸들링은 편하다고 말하기 어려운데, 넘파이 기반으로 많은 부분이 작성된 판다스는 넘파이보다 훨씬 유연하고 편리하게 데이터 핸들링을 가능케 한다. 판다스는 파이썬의 리스트, 컬렉션, 넘파이 등의 내부 데이터뿐만 아니라 CSV 등의 파일을 쉽게 DataFrame으로 변경해 데이터의 가공 및 분석을 편리하게 수행할 수 있게 만들어준다.
*   판다스의 핵심 객체는 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체인 DataFrame이다. 그 전에, 다른 중요 객체인 Index와 Series에 대해 먼저 살펴볼 필요가 있다. Index는 RDBMS의 PK처럼 개별 데이터를 고유하게 식별하는 Key 값이다. Series와 DataFrame은 모두 Index를 Key 값으로 가지고 있는데, Series와 DataFrame의 가장 큰 차이는 Series는 칼럼이 하나 뿐인 데이터 구조체이고, DataFrame은 칼럼이 여러 개인 데이터 구조체라는 점이다. DataFrame은 여러 개의 Series로 이루어졌다고 할 수 있다. 판다스는 csv, tab과 같은 다양한 유형의 분리 문자로 칼럼을 분리한 파일을 손쉽게 DataFrame으로 로딩할 수 있게 해준다.

### 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API






In [2]:
import pandas as pd



*   pandas를 pd로 에일리어스(alias)해 임포트하는 것이 관례이다.
*   판다스는 다양한 포맷으로 된 파일을 DataFrame으로 로딩할 수 있는 편리한 API를 제공한다. 대표적으로 read_csv(), read_table(), read_fwf()가 있다. read_csv()는 이름에서도 알 수 있듯 CSV(칼럼을 ','로 구부한 파일 포맷) 파일 포맷 변환을 위한 API입니다. read_table()과 read_csv()의 가장 큰 차이는 필드 구분 문자(Delimeter)가 콤마(')냐, 탭('\t')이냐의 차이이다. read_table()의 디폴트 필드 구분 문자는 탭 문자이다. read_csv()는 CSV뿐만 아니라 어떤 필드 구분 문자 기반의 파일 포맷도 DataFrame으로 변환이 가능한데, read_csv()의 인자인 sep에 해당 구분 문자를 입력하면 된다. 가령 탭으로 필드가 구분되어 있다면 read_csv('파일명', sep= '\t')처럼 쓰면 된다. read_csv()에서 sep 인자를 생략하면 자동으로 콤마로 할당한다. (sep = ','). read_csv()와 read_table()은 기능상 큰 차이가 없다. read_fwf()는 Fixed Width, 즉 고정 길이 기반의 칼럼 포맷을 DataFrame으로 로딩하기 위한 API이다.


*   read_csv(filepath_or_buffer, sep= ',',...) 함수에서 가장 중요한 인자는 filepath이다. 나머지 인자는 지정하지 않으면 디폴트 값으로 할당된다. Filepath에는 로드하려는 데이터 파일의 경로를 포함한 파일명을 입력하면 된다.파일명만 입력되면 파이썬 실행 파일이 있는 디렉터리와 동일한 디렉터리에 있는 파일명을 로딩한다.  





In [4]:
titanic_df = pd.read_csv('/content/titanic_train.csv')
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [5]:
titanic_df = pd.read_csv('/content/titanic_train.csv')
print('titanic 변수 type', type(titanic_df))
titanic_df

titanic 변수 type <class 'pandas.core.frame.DataFrame'>


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,111369,30.0000,C148,C




*   pd.read_csv()는 호출 시 파일명 인자로 들어온 파일을 로딩해 DataFrame 객체로 반환한다. read_csv()는 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지하고 칼럼으로 변환한며, 콤마로 분리된 데이터들이 해당 칼럼에 맞게 할당된다.
*   맨 왼쪽을 보면 파일에 기재돼 있지 않는 데이터값이 로우 순으로 0, 1, 2, 3......과 같이 순차적으로 표시된다. 칼럼명도 표시되지 않는 이 데이터들은 바로 판다스의 Index 객체 값으로, 모든 DataFrame 내의 데이터는 생성되는 순간 고유의 Index 값을 가지게 된다. 즉, Index는 RDBMS의 PK와 유사하게 고유의 레코드를 식별하는 역할을 한다는 것이다.
*   DataFrame.head()는 DataFrame의 맨 앞에 있는 N개의 로우를 반환한다. head(3)은 맨 앞 3개의 로우를 반환하며, Default는 5개이다.


In [6]:
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


In [20]:
print('DataFrame 크기: ', titanic_df.shape)

DataFrame 크기:  (891, 12)




*   DataFrame의 행과 열 크기를 알아보는 가장 좋은 방법은 생성된 DataFrame 객체의 shape 변수를 이용하는 것이다. shape는 DataFrame의 행과 열을 튜플 형태로 반환한다.



In [7]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


*   DataFrame은 데이터뿐만 아니라 칼럼의 타입, Null 데이터 개수, 데이터 분포도 등의 메타 데이터 등도 조회가 가능하다. 이를 위한 대표적인 메서드로 info()와 describe()가 있다. info() 메서드를 통해서 총 데이터 건수, 데이터 타입, Null 건수를 알 수 있다. descirbe() 메서드는 칼럼별 숫자형 데이터값의 n-percentile 분포도, 평균값, 최댓값, 최솟값을 나타낸다. describe() 메서드는 오직 숫자형(int, float 등) 칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외시킨다. describe() 메서드만으로 정확한 분포도를 알기는 무리지만, 대략적인 수준의 분포도를 확인할 수 있어 머신러닝 알고리즘 성능을 향상시키는 데 기여할 수 있다.

*   count는 Not Null인 데이터 건수, mean은 전체 데이터의 평균값, std는 표준편차이다. 사분위값들은 describe()는 해당 숫자 칼럼이 숫자형 카테고리 칼럼인지를 판단할 수 있게 도와준다. 카테고리 칼럼은 특정 범주에 속하는 값을 코드화한 칼럼이고, 이를 숫자로 표시할 수 있다.
*   DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 Series 형태로 특정 칼럼 데이터 세트가 반환된다. 이렇게 반환된 Series 객체에 value_counts() 메서드를 호출하면 해당 칼럼값의 유형과 건수를 확인할 수 있다. value_counts()는 지정된 칼럼의 데이터값 건수를 반환한다. value_counts()는 데이터의 분포도를 확인하는 데 매우 유용한 함수이며, 많은 건수 순서대로 정렬해 값을 반환한다. DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환한다.



In [8]:
titanic_df.describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [9]:
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)

Pclass
3    491
1    216
2    184
Name: count, dtype: int64


In [10]:
titanic_pclass = titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>




*   Series는 Index와 단 하나의 칼럼으로 구성된 데이터 세트이다. 한 개 칼럼의 데이터만 출력되지 않고 맨 왼쪽에 0부터 시작하는 순차 값이 있는데, 이것은 DataFrame의 인덱스와 동일한 인덱스이다. 오른쪽은 Series의 해당 칼럼의 데이터값이다. 모든 Series와 DataFrame은 인덱스를 반드시 가진다.
*   value_counts() 메서드는 단일 칼럼으로 되어 있는 Series 객체에서 호출하는 것이 칼럼별 데이터 값의 분포도를 더 명시적으로 파악하기 쉽다. DataFrame 객체에서도 물론 호출할 수 있다.



In [11]:
titanic_pclass.head()

,Pclass
0,3
1,1
2,3
3,1
4,3


In [13]:
value_counts = titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
Pclass
3    491
1    216
2    184
Name: count, dtype: int64




*   value_counts()가 반환하는 데이터 타입 역시 Series 객체이다. 반환된 Series 객체의 값을 보면 맨 왼쪽이 인덱스값이며, 오른쪽이 데이터값인데, 이번에는 인덱스가 단순히 0부터 시작하는 순차 값이 아니다. 이처럼 인덱스는 단순히 순차 값과 같은 의미 없는 식별자만 할당하는 것이 아니라 고유성이 보장된다면 의미 있는 데이터값 할당도 가능하다. value_counts()는 칼럼 값별 데이터 건수를 반환하므로 고유 칼럼 값을 식별자로 사용할 수 있다.
*   인덱스는 DataFrame, Series가 만들어진 이후에도 변경할 수 있으며, 숫자형뿐만 아니라 문자열도 가능하다. 단, 모든 인덱스는 고유성이 보장되어야 한다.
*   value_counts() 메서드를 사용할 때에는 Null 값을 무시하고 결과값을 내놓기 쉽다는 점을 유의해야 한다. value_counts()는 Null 값을 포함하여 개별 데이터 값의 건수를 계산할지를 dropna 인자로 판단한다. dropna의 기본값은 True이며, Null 값을 무시하고 개별 데이터 값의 건수를 계산한다. 만약 Null 값을 포함해 value_counts()를 적용하고자 한다면 value_counts(dropna = False)와 같이 dropna 인자값을 False로 입력해주면 된다.




In [14]:
print('titanic_df 데이터 건수: ', titanic_df.shape[0])
print('기본 설정인 dropna = True로 value_counts()')
# value_counts()는 디폴트로 dropna = True이므로 value_counts(dropna = True)와 동일.
print(titanic_df['Embarked'].value_counts())
print(titanic_df['Embarked'].value_counts(dropna = False))

titanic_df 데이터 건수:  891
기본 설정인 dropna = True로 value_counts()
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64
Embarked
S      644
C      168
Q       77
NaN      2
Name: count, dtype: int64


value_counts()를 이용해 타이타닉 데이터 세트의 Embarked 칼럼의 데이터 값 분포를 살표보면, 데이터 세트의 전체 건수는 891건이며 이는 titanic_df.shape[0]로 확인할 수 있다. Embarked 칼럼은 전체 891건 중 2건의 데이터가 Null이다. titanic_df['Embarked'].value_counts()의 결과는 'S', 'C', 'Q' 값으로 각각 644, 168, 77을 반환한다. 전체를 합하면 889가 되기에 2건의 Null 데이터를 제외하고 연산을 수행했음을 알 수 있다.

### DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환
*   기본적으로 DataFrame은 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있다. 특히 사이킷런의 많은 API는 DataFrame을 인자로 입력받을 수 있지만, 기본적으로 넘파이 ndarray를 입력 인자로 사용하는 경우가 대부분이다. 따라서 DataFrame과 넘파이 ndarray 상호 간의 변환은 매우 빈번하게 발생한다.
*   넘파이 ndarray, 파이썬 리스트, 딕셔너리로부터 DataFrame을 생성한다. DataFrame은 리스트와 넘파이 ndarray와는 다르게 칼럼명을 가지고 있어 보다 편하게 데이터 핸들링이 가능하다. 일반적으로 DataFrame으로 변환 시 이 칼럼명을 지정해주며, 지정하지 않으면 자동으로 칼럼명을 할당한다. 판다스 DataFrame 객체의 생성 인자 data는 리스트나 딕셔너리 또는 넘파이 ndarray를 입력받고, 생성 인자 columns는 칼럼명 리스트를 입력받아서 쉽게 DataFrame을 생성할 수 있다.
*   DataFrame은 기본적으로 행과 열을 가지는 2차원 데이터이므로 2차원 이하의 데이터들만 DataFrame으로 변환될 수 있다.





In [16]:
import numpy as np

col_name1 = ['col1']
list1 = [1, 2, 3]
array1 = np.array(list1)
print('array1 shape: ', array1.shape)
# 리스트를 이용해 DataFrame 생성
df_list1 = pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame: \n', df_list1)
# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1 = pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame: \n', df_array1)

array1 shape:  (3,)
1차원 리스트로 만든 DataFrame: 
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame: 
    col1
0     1
1     2
2     3


1차원 형태의 리스트와 넘파이 ndarray를 DataFrame으로 변환한 결과이다. 1차원 데이터이므로 칼럼은 1개만 필요하다.

In [17]:
# 3개의 칼럼명이 필요함
col_name2 = ['col1', 'col2', 'col3']

# 2행 x 3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2 = [[1, 2, 3],
         [11, 12, 13]]
array2 = np.array(list2)
print('array2 shape:', array2.shape)
df_list2 = pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame: \n', df_list2)
df_array2 = pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame: \n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    12    13


2차원 형태의 데이터를 기반으로 DataFrame을 생성한 결과이다. 2행 3열 형태의 리스트와 ndarray를 기반으로 DataFrame을 생성하므로 칼럼명은 3개가 필요하다.

In [20]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict = {'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}
df_dict = pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame: \n', df_dict)

딕셔너리로 만든 DataFrame: 
    col1  col2  col3
0     1     2     3
1    11    22    33


딕셔너리를 DataFrame으로 변환한 결과이다. Key 값은 칼럼명, 각 Value는 각 칼럼 데이터로 매핑되었음을 알 수 있다.

*   일반적으로 딕셔너리를 DataFrame으로 변환 시에는 딕셔너리의 키(Key)는 칼럼명으로, 딕셔너리의 값(Value)는 키에 해당하는 칼럼 데이터로 변환된다. 따라서 키의 경우에는 문자열, 값의 경우에는 리스트(또는 ndarray) 형태로 딕셔너리를 구성한다.

*   DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환할 수도 있다. 많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용하므로, 데이터 핸들링은 DataFrame을 이용하더라도 머신러닝 패키지의 입력 인자 등에 적용하기 위해 다시 넘파이 ndarray로 변환하는 경우가 빈번하게 발생한다. DataFrame을 넘파이 ndarray로 변환하는 것은 DataFrame 객체의 values를 이용해 쉽게 할 수 있다. DataFrame을 리스트로 변환할 때에는 values로 얻은 ndarry에 tolist()를 호출하면 된다. 딕셔너리로의 변환은 DataFrame 객체의 to_dict() 메서드를 호출하는데, 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환된다.



In [22]:
# DataFrame을 ndarray로 변환
array3 = df_dict.values
print('df_dict.values 타입:', type(array3), 'df_dict.values shape: ', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape:  (2, 3)
[[ 1  2  3]
 [11 22 33]]


In [24]:
#DataFrame을 리스트로 변환
list3 = df_dict.values.tolist()
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

#DataFrame을 딕셔너리로 변환
dict3 = df_dict.to_dict('list')
print('\n df_dict.to_list() 타입:', type(dict3))  #이 부분 질문할 것
print(dict3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_list() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}




### DataFrame의 칼럼 데이터 세트 생성과 수정
*   DataFrame의 칼럼 데이터 세트 생성과 수정 역시 [] 연산자를 이용해 쉽게 할 수 있다. 먼저 Titanic DataFrame의 새로운 칼럼 Age_0을 추가하고 일괄적으로 0 값을 할당한 후, DataFrame [] 내에 새로운 칼럼명을 입력하고 값을 할당해주기만 하면 된다.
*   Series에 상숫값을 할당하면 (e.g titanic_df['Age_0'] = 0) Series의 모든 데이터 세트에 일괄적으로 적용된다.
*   기존 칼럼 Series의 데이터를 이용해 새로운 칼럼 Series를 만들 수도 있다.
*   DataFrame 내의 기존 칼럼 값도 일괄적으로 업데이트할 수 있다. 업데이트를 원하는 칼럼 Series를 DataFrame[] 내에 칼럼명으로 입력한 뒤에 값을 할당해주면 된다.





In [25]:
titanic_df['Age_0'] = 0
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0


새로운 칼럼명 'Age_0'으로 모든 데이터값이 0으로 할당된 Series가 기존 DataFrame에 추가됨을 알 수 이다.

In [26]:
titanic_df['Age_by_10'] = titanic_df['Age'] * 10
titanic_df['Family_No'] = titanic_df['SibSp'] + titanic_df['Parch'] + 1
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,220.0,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,380.0,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,260.0,1


기존 칼럼 Series의 데이터를 이용해 새로운 칼럼 Series를 만든 결과이다. 기존 칼럼 Series를 가공해 새로운 칼럼 Series인 Age_by_10과 Family_No가 새롭게 DataFrame에 추가된 것을 알 수 있다.

In [27]:
titanic_df['Age_by_10'] = titanic_df['Age_by_10'] + 100
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,320.0,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,480.0,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,360.0,1


DataFrame 내의 기존 칼럼 값을 일괄적으로 업데이트한 결과이다.

### DataFrame 데이터 삭제


*   DataFrame에서 데이터의 삭제는 drop() 메서드를 이용한다. drop() 메서드의 원형은 다음과 같다.


> DataFrame.drop(labels=None, axis=0, index=None, columns=None, level=None, inplace=False, errors='raise')


*   이 중 가장 중요한 파라미터는 labels, axis, inplace이다. 먼저 axis 값에 따라서 특정 칼럼 또는 특정 행을 드랍한다. axis 0은 로우 방향 축, axis 1은 칼럼 방향 축이며 판다스의 DataFrame은 2차원 데이터만 다루므로 두 axis로만 axis가 구성되어 있다. 따라서 drop() 메서드에 axis=1을 입력하면 칼럼 축 방향으로 드롭을 수행하므로 칼럼을 드랍하겠다는 것이다. labels에 원하는 칼럼명을 입력하고 axis=1을 입력하면 지정된 칼럼을 드롭한다. 또, drop() 메서드에 axis=0을 입력하면 로우 축 방향으로 드롭을 수행하므로 특정 로우를 드롭하겠다는 것이다. DataFrame의 특정 로우를 가리키는 것은 인덱스이므로, axis를 0으로 지정하면 DataFrame은 자동으로 labels에 오는 값을 인덱스로 간주한다.

*   이 책에서 drop() 메서드가 사용되는 대부분의 경우는 칼럼을 드롭하는 경우다. 기존 칼럼 값을 가공해 새로운 칼럼을 만들고 삭제하는 경우가 많아 axis=1로 설정하고 드롭하는 경우가 많을 수밖에 없다. axis=0으로 설정하고 로우 레벨로 삭제를 하는 경우는 이상치 데이터를 삭제하는 경우에 주로 사용된다.
*   Inplace는 디폴트 값이 False이므로 inplace 파라미터를 기재하지 않으면 자동으로 False가 된다. Inplace=False이면 자기 자신의 DataFrame의 데이터를 삭제하지 않고, 삭제된 결과 DataFrame을 반환한다. inplace=True로 설정하면 자신의 DataFrame의 데이터를 삭제한다. 또, 여러 개의 칼럼을 삭제하고 싶으면 리스트 형태로 삭제하고자 하는 칼럼명을 입력해 labels 파라미터로 입력하면 된다.


*   한 가지 유의할 점은 drop() 시 inplace=True로 설정하면 반환 값이 None(아무 값도 아님)이 된다는 점이다. 그러므로 inplace=True로 설정한 채로 반환 값을 다시 자신의 DataFrame 객체로 할당하면 안 된다.





In [30]:
titanic_drop_df = titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,320.0,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,480.0,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,360.0,1


In [31]:
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,Age_0,Age_by_10,Family_No
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S,0,320.0,2
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C,0,480.0,2
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S,0,360.0,1


원본 Titanic DataFrame을 확인해보면 삭제되었다고 생각했던 'Age_0' 칼럼이 여전히 존재한다. 그 이유는 앞의 예제 코드에서 inplace = False라고 설정했기 때문이다. titanic_df DataFrame 객체 변수는 'Age_0' 칼럼이 삭제되지 않았으며 titanic_df.drop('Age_0', axis=1)의 수행 결과로 반환된 titanic_drop_df DataFrame 객체 변수만 'Age_0' 칼럼이 삭제되었다.

In [32]:
drop_result = titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print('inplace=True로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True로 drop 후 반환된 값: None


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S


titanic_df의 'Age_0, 'Age_by_10', 'Family_No' 칼럼을 모두 삭제한 결과이다.

In [33]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))

titanic_df.drop([0, 1, 2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


axis=0으로 설정해 index 0, 1, 2(맨 앞 3개 데이터) 로우를 삭제해본 결과이다.삭제 후에 head(3)으로 확인한 맨 앞 3개 데이터의 index는 3, 4, 5이므로, 인덱스 0, 1, 2에 위치한 로우가 삭제되었음을 알 수 있다.

### Index 객체



*   판다스의 Index 객체는 DataFrame, Series의 레코드를 고유하게 식별하는 객체로, RDBMS의 PK(Primary Key)와 유사한 기능을 한다. DataFrame, Series에서 Index 객체만 추출하려면 DataFrame.index 또는 Series.index 속성을 통해 가능하다. 반환된 Index 객체의 실제 값은 넘파이 1차원 ndarray로 볼 수 있다. Index 객체의 values 속성으로 ndarray 값을 알 수 있다.
*   Index 객체는 식별성 데이터를 1차원 array로 가지고 있으며, ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능하다. 그러나 한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없다.




In [34]:
# 원본 파일 다시 로딩 : drop으로 원본 Titanic DataFrame의 일부 데이터가 삭제되었으므로 다시 csv 파일에서 로딩 후 수행.
titanic_df = pd.read_csv('/content/titanic_train.csv')
# Index 객체 추출
indexes = titanic_df.index
print(indexes)
# Index 객체를 실제 값 array로 변환
print('Index 객체 array 값: \n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array 값: 
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 2

In [35]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


In [36]:
indexes[0] = 5

TypeError: Index does not support mutable operations



*   Series 객체는 Index 객체를 포함하지만, Series 객체에 연산 함수를 적용할 때 Index는 연산에서 제외된다. Index는 오직 식별용으로만 사용된다.


In [39]:
series_fair = titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n', (series_fair + 3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.9493
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64




*   DataFrame 및 Series에 reset_index() 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 'index'라는 새로운 칼럼명으로 추가한다. reset_index()는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우, 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용한다.
*   Series에 reset_index()를 적용하면 Series가 아닌 DataFrame이 반환되니 유의해야 한다. 기존 인덱스가 칼럼으로 추가돼 칼럼이 2개가 되므로, Series가 아닌 DataFrame이 반환되는 것이다. reset_index()의 parameter 중 drop=True로 설정하면, 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제, 즉 Drop된다. 새로운 칼럼이 추가되지 않으므로 그대로 Series로 유지된다.


In [40]:
titanic_reset_df = titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

,index,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,1,2,1,1,"Cumings, Mr...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,2,3,1,3,"Heikkinen, ...",female,26.0,0,0,STON/O2. 31...,7.9250,NaN,S


In [41]:
print('### before reset_index ###')
value_counts = titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입: ', type(value_counts))
new_value_counts = value_counts.reset_index(inplace=False)
print('### after reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
Pclass
3    491
1    216
2    184
Name: count, dtype: int64
value_counts 객체 변수 타입:  <class 'pandas.core.series.Series'>
### after reset_index ###
   Pclass  count
0       3    491
1       1    216
2       2    184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


Series와 value_counts()를 소개하는 예제에서 'Pclass' 칼럼 Series의 value_counts()를 수행하면, 'Pclass' 고유 값이 식별자 인덱스 역할을 했다. 그러나 이보다는 연속 숫자형 인덱스가 고유 식별자로 더 적합해보이므로, reset_index()를 이용해 인덱스를 다시 만들었다.

### 데이터 셀렉션 및 필터링


*   넘파이의 데이터 핸들링은 데이터 분석용으로 사용하기에는 편의성이 떨어지기에, 판다스는 이를 개선하려고 노력하는 과정에서 넘파이 기능을 계승하기도, 때로는 완전히 다른 기능을 제공하기도 한다. 판다스의 DataFrame과 Series끼리도 이러한 데이터 셀렉션 기능이 달라지는 부분이 있어 주의가 필요하다.
*   넘파이의 경우 '[]' 연산자 내 단일 값 추출, 슬라이싱, 팬시 인덱싱, 불린 인덱싱을 통해 데이터를 추출했다. 판다스의 경우 iloc[], loc[] 연산자를 통해 동일한 작업을 수행한다.


*   넘파이와 DataFrame 간 데이터 셀렉션에서 가장 유의해야 할 부분은 '[]' 연산자이다. 넘파이에서 [] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있었다. 하지만 DataFrame 바로 뒤에 있는 '[]' 안에 들어 갈 수 있는 것은 칼럼명 문자(또는 칼럼명의 리스트 객체), 또는 인덱스로 변환 가능한 표현식이다. DataFrame 뒤에 있는 []는 칼럼만 지정할 수 있는 '칼럼 지정 연산자'로 이해하면 된다.
*   판다스의 인덱스 형태로 변환 가능한 표현식은 [] 내에 입력할 수 있다.





In [42]:
print('단일 칼럼 데이터 추출: \n', titanic_df['Pclass'].head(3))
print('\n 여러 칼럼의 데이터 추출: \n', titanic_df[ ['Survived', 'Pclass']]. head(3))
print('[] 안에 숫자 index는 KeyError 오류 발생: \n', titanic_df[0])

단일 칼럼 데이터 추출: 
 0    3
1    1
2    3
Name: Pclass, dtype: int64

 여러 칼럼의 데이터 추출: 
    Survived  Pclass
0         0       3
1         1       1
2         1       3


KeyError: 0



*   슬라이싱과 불린 인덱싱 표현을 통해 원하는 데이터를 추출할 수 있으나, 불린 인덱싱은 매우 자주 사용되는 반면 슬라이싱 연산은 지양된다.



In [43]:
titanic_df[0:2]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mr...",female,38.0,1,0,PC 17599,71.2833,C85,C


In [45]:
titanic_df[titanic_df['Pclass'] == 3].head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.250,NaN,S
2,3,1,3,"Heikkinen, ...",female,26.0,0,0,STON/O2. 31...,7.925,NaN,S
4,5,0,3,"Allen, Mr. ...",male,35.0,0,0,373450,8.050,NaN,S




*   판다스는 DataFrame의 로우나 칼럼을 지정하여 데이터를 선택할 수 있는 인덱싱 방식으로 iloc[]와 loc[]를 제공한다. iloc[]는 위치(Location) 기반 인덱싱 방식으로 동작하며. loc[]는 명칭(Label) 기반 인덱싱 방식으로 동작한다. 위치 기반 인덱싱은 행과 열 위치를, 0을 출발점으로 하는 세로축, 가로축 좌표 정숫값으로 지정하는 방식이다. 명칭 기반 인덱싱은 DataFrame의 인덱스 값으로 행 위치를, 칼럼의 명칭으로 열 위치를 지정하는 방식이다.
*   iloc[]는 위치 기반 인덱싱만 허용되기 때문에 행과 열의 좌표 위치에 해당하는 값으로 정숫값 또는 정수형의 슬라이싱, 팬시 리스트 값을 입력해줘야 한다.



In [46]:
# DataFrame 생성
data = {'Name' : ['Chulmin','Eunkyung', 'Jinwoong', 'Soobeom'],
        'Year' : [2011, 2016, 2015, 2015],
        'Gender' : ['Male', 'Female', 'Male', 'Male']
        }
data_df = pd.DataFrame(data, index=['one', 'two', 'three', 'four'])
data_df

,Name,Year,Gender
one,Chulmin,2011,Male
two,Eunkyung,2016,Female
three,Jinwoong,2015,Male
four,Soobeom,2015,Male


생성된 DataFrame은 인덱스 값으로 숫자값이 아닌 문자열로 'one', 'two', 'three', 'four'를 가지고 있다.



*   iloc[]를 이용해 특정 행, 특정 열의 데이터를 추출할 때에는 data_df.iloc[0, 0]과 같이 행과 열 위치에 위치 기반 인덱싱 값을 입력하면 된다. 반면 iloc[]에 DataFrame의 인덱스 값이나 칼럼명을 입력하면 오류가 발생한다.




In [47]:
data_df.iloc[0, 0]

'Chulmin'

In [48]:
# 아래 코드는 오류를 발생시킨다
data_df.iloc[0, 'Name']

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

In [49]:
# 아래 코드는 오류를 발생시킨다
data_df.iloc['one', 0]

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

In [51]:
# 두번째 행의 첫번째 열 위치에 있는 단일 값 변환
data_df.iloc[1, 0]

'Eunkyung'

In [52]:
# 세번째 행의 두번째 열 위치에 있는 단일 값 변환
data_df.iloc[2, 1]

np.int64(2015)

In [54]:
# 0:2 슬라이싱 범위의 첫번째에서 두번째 행과 첫번째, 두번째 열에 해당하는 DataFrame 변환
data_df.iloc[0:2, [0,1]]

,Name,Year
one,Chulmin,2011
two,Eunkyung,2016


In [55]:
# 0:2 슬라이싱 범위의 첫번째에서 두번째 행의 0:3 슬라이싱 범위의 첫번째에서 세번째 열 범위에 해당하는 DataFrame 반환
data_df.iloc[0:2, 0:3]

,Name,Year,Gender
one,Chulmin,2011,Male
two,Eunkyung,2016,Female


In [56]:
# 전체 DataFrame 반환
data_df.iloc[:]

,Name,Year,Gender
one,Chulmin,2011,Male
two,Eunkyung,2016,Female
three,Jinwoong,2015,Male
four,Soobeom,2015,Male




*   iloc[]는 열 위치에 -1을 입력해 DataFrame의 가장 마지막 열 데이터를 가져오는 데 자주 사용한다. 넘파이와 마찬가지로 판다스의 인덱싱에서도 -1은 맨 마지막 데이터 값을 의미한다. 특히 머신 러닝 학습 데이터의 맨 마지막 칼럼이 타깃 값이 경우가 많은데, 이 경우 iloc[:, -1]을 하면 맨 마지막 칼럼의 값, 즉 타깃 값을 가져오고, iloc[:, :-1]을 하게 되면 처음부터 맨 마지막 칼럼을 제외한 모든 칼럼의 값, 즉 피처값들을 가져오게 된다.
*   iloc[]는 슬라이싱과 팬시 인덱싱을 제공한다. 그러나 명확한 위치 기반 인덱싱이 사용되어야 하는 제약 때문에 불린 인덱싱은 제공하지 않는다.


In [57]:
print("\n 맨 마지막 칼럼 데이터 [:, -1] \n", data_df.iloc[:, -1])
print("\n 맨 마지막 칼럼을 제외한 모든 데이터 [:, :-1] \n", data_df.iloc[:, :-1])


 맨 마지막 칼럼 데이터 [:, -1] 
 one        Male
two      Female
three      Male
four       Male
Name: Gender, dtype: object

 맨 마지막 칼럼을 제외한 모든 데이터 [:, :-1] 
            Name  Year
one     Chulmin  2011
two    Eunkyung  2016
three  Jinwoong  2015
four    Soobeom  2015




*   loc[]는 명칭(Label) 기반으로 데이터를 추출한다. 행 위치에는 DataFrame 인덱스 값을, 그리고 열 위치에는 칼럼명을 입력해 loc[인덱스값, 칼럼명]와 같은 형식으로 데이터를 추출할 수 있다.
*   일반적으로 인덱스는 0부터 시작하는 정숫값인 경우가 많기 때문에 실제 인덱스 값을 확인하지 않고 loc[]의 행 위치에 무턱대고 정숫값을 입력해서 오류가 나는 경우가 종종 발생하므로 유의해야 한다.


In [58]:
data_df.loc['one', 'Name']

'Chulmin'

In [59]:
# 다음 코드는 오류를 발생시킨다
data_df.loc[0, 'Name']

KeyError: 0



*   loc[]에 슬라이싱 기호 ':'를 적용할 때 한 가지 유의할 점이 있다. 일반적으로 슬라이싱을 '시작 값: 종료 값'과 같이 지정하면 시작 값 ~ 종료 값-1까지의 범위를 의미한다. 그러나 loc[]에 슬라이싱 기호를 적용하면 종료 값-1이 아니라 종료 값까지 포함하는 것을 의미하는데, 이는 명칭 기반 인덱싱의 특성 때문이다. 명칭은 숫자형이 아닐 수 있기 때문에 -1을 할 수가 없다.
*   loc[]는 iloc[]와 달리 불린 인덱싱이 가능하다.


In [60]:
print('위치기반 iloc slicing \n', data_df.iloc[0:1, 0], '\n')
print('명칭기반 loc slicing \n', data_df.loc['one':'two', 'Name'])

위치기반 iloc slicing 
 one    Chulmin
Name: Name, dtype: object 

명칭기반 loc slicing 
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


위치 기반 iloc[0:1]은 0번째 행 위치에 해당하는 1개의 행을 반환했지만, 명칭 기반 loc['one':'two']는 2개의 행을 반환했다.

In [61]:
# 인덱스 값 three인 행의 Name 칼럼의 단일값 반환
data_df.loc['three', 'Name']

'Jinwoong'

In [62]:
# 인덱스 값 one부터 two까지 행의 Name과 Year 칼럼에 해당하는 DataFrame 반환
data_df.loc['one':'two', 'Name':'Year']

,Name,Year
one,Chulmin,2011
two,Eunkyung,2016


In [63]:
# 인덱스 값 one부터 three까지의 행의 Name부터 Gender 칼럼까지의 DataFrame 반환
data_df.loc['one':'three', 'Name':'Gender']

,Name,Year,Gender
one,Chulmin,2011,Male
two,Eunkyung,2016,Female
three,Jinwoong,2015,Male


In [64]:
# 모든 데이터 값 반환
data_df.loc[:]

,Name,Year,Gender
one,Chulmin,2011,Male
two,Eunkyung,2016,Female
three,Jinwoong,2015,Male
four,Soobeom,2015,Male


In [65]:
# 불린 인덱싱을 이용해 Year 칼러의 값이 2014 이상인 모든 데이터를 추출
data_df.loc[data_df.Year >= 2014]

,Name,Year,Gender
two,Eunkyung,2016,Female
three,Jinwoong,2015,Male
four,Soobeom,2015,Male




*   불린 인덱싱은 매우 편리한 데이터 필터링 방식이기에 오히려 iloc나 loc와 같이 명확히 인덱싱을 지정하는 방식보다 더 선호되는 경우가 많다. 명칭이나 위치 지정 인덱싱에서 가져올 값은 주로 로직이나 조건에 의해 계산된 뒤 행 위치, 열 위치 값으로 입력되는데, 불린 인덱싱은 그럴 필요 없이 처음부터 가져올 값을 조건으로 [] 내에 입력하는 방식이기에 자동으로 원하는 값을 필터링할 수 있기 때문이다. 그리고 불린 인덱싱은 [], loc[]에서 공통으로 지원한다. 단지 iloc[]는 정수형 값이 아닌 불린 값에 대해서는 지원하지 않기 때문에 불린 인덱싱이 지원되지 않는다.
*   [] 연산자 내에 불린 조건을 입력하면 불린 인덱싱으로 자동으로 결과를 찾을 수 있다. 이 때 반환되는 객체는 DataFrame이므로 원하는 칼럼명만 별도로 추출할 수도 있다. loc[]을 이용해도 동일하게 적용할 수 있다.
*   여러 개의 복합 조건도 결합해 적용할 수 있다. (i) and 조건일 때는 & (ii) or 조건일 때는 | (iii) Not 조건일 때는 ~ 이며, 개별 조건은 ()로 묶고 복합 조건 연산자를 사용하면 된다.
*   개별 조건을 변수에 할당하고 이들 변수를 결합해서 불린 인덱싱을 수행할 수도 있다.





In [66]:
titanic_df = pd.read_csv('/content/titanic_train.csv')
titanic_boolean = titanic_df[titanic_df['Age'] > 60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
33,34,0,2,"Wheadon, Mr...",male,66.0,0,0,C.A. 24579,10.5000,NaN,S
54,55,0,1,"Ostby, Mr. ...",male,65.0,0,1,113509,61.9792,B30,C
96,97,0,1,Goldschmidt...,male,71.0,0,0,PC 17754,34.6542,A5,C
116,117,0,3,"Connors, Mr...",male,70.5,0,0,370369,7.7500,NaN,Q
170,171,0,1,Van der hoe...,male,61.0,0,0,111240,33.5000,B19,S
252,253,0,1,"Stead, Mr. ...",male,62.0,0,0,113514,26.5500,C87,S
275,276,1,1,"Andrews, Mi...",female,63.0,1,0,13502,77.9583,D7,S
280,281,0,3,"Duane, Mr. ...",male,65.0,0,0,336439,7.7500,NaN,Q
326,327,0,3,"Nysveen, Mr...",male,61.0,0,0,345364,6.2375,NaN,S
438,439,0,1,"Fortune, Mr...",male,64.0,1,4,19950,263.0000,C23 C25 C27,S


In [67]:
titanic_df[titanic_df['Age']>60][['Name', 'Age']].head(3) # 칼럼이 두 개 이상이므로 [[]] 사용

,Name,Age
33,"Wheadon, Mr...",66.0
54,"Ostby, Mr. ...",65.0
96,Goldschmidt...,71.0


In [69]:
titanic_df.loc[titanic_df['Age']>60, ['Name', 'Age']].head(3) # 이 때 ['Name', 'Age']는 칼럼 위치에 놓여야 한다

,Name,Age
33,"Wheadon, Mr...",66.0
54,"Ostby, Mr. ...",65.0
96,Goldschmidt...,71.0


In [70]:
titanic_df[(titanic_df['Age'] > 60) & (titanic_df['Pclass'] == 1) & (titanic_df['Sex'] == 'female')]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
275,276,1,1,"Andrews, Mi...",female,63.0,1,0,13502,77.9583,D7,S
829,830,1,1,"Stone, Mrs....",female,62.0,0,0,113572,80.0000,B28,NaN


In [71]:
cond1 = titanic_df['Age'] > 60
cond2 = titanic_df['Pclass']==1
cond3 = titanic_df['Sex']=='female'
titanic_df[cond1 & cond2 & cond3]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
275,276,1,1,"Andrews, Mi...",female,63.0,1,0,13502,77.9583,D7,S
829,830,1,1,"Stone, Mrs....",female,62.0,0,0,113572,80.0000,B28,NaN


### 정렬, Aggregation 함수, GroupBy 적용



*   DataFrame과 Series의 정렬을 위해서는 sort_values() 메서드를 이용한다. (RDBMS SQL의 order by 키워드와 매우 유사하다) sort_values()의 주요 입력 파라미터는 by, ascending, inplace이다. by로 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행한다. ascending=True로 설정하면 오름차순으로 정렬하며, ascending=False로 설정하면 내림차순으로 정렬한다. 기본은 ascending=True이다. inplace=False로 설정하면 sort_values()를 호출한 DataFrame은 그대로 유지하며 정렬된 DataFrame을 결과로 반환한다. inplae=True로 설정하면 호출한 DataFrame의 정렬 결과를 그대로 적용한다. 기본은 inplace=False이다.
*   여러 개의 칼럼으로 정렬하려면 by에 리스트 형식으로 정렬하려는 칼럼을 입력하면 된다.



In [73]:
titanic_sorted = titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
845,846,0,3,"Abbing, Mr....",male,42.0,0,0,C.A. 5547,7.55,NaN,S
746,747,0,3,"Abbott, Mr....",male,16.0,1,1,C.A. 2673,20.25,NaN,S
279,280,1,3,"Abbott, Mrs...",female,35.0,1,1,C.A. 2673,20.25,NaN,S


In [74]:
titanic_sorted = titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
868,869,0,3,van Melkebe...,male,NaN,0,0,345777,9.5,NaN,S
153,154,0,3,van Billiar...,male,40.5,0,2,A/5. 851,14.5,NaN,S
282,283,0,3,de Pelsmaek...,male,16.0,0,0,345778,9.5,NaN,S




*   DataFrame에서 min(), max(), sum(), count()와 같은 aggregation 함수의 적용은 RDBMS SQL의 aggregtion 함수 적용과 유사하다. 다만 DataFrame의 경우, DataFrame에서 바로 aggregation을 호출할 경우 모든 칼럼에 해당 aggregation을 적용한다는 차이가 있다. 단, count()는 Null 값을 반영하지 않은 결과를 반환해 Null 값이 있는 Age, Cabin, Embarked 칼럼은 count()의 결과 값이 다르다.
*   특정 칼럼에 aggregation 함수를 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 aggregation을 적용하면 된다. SQL의 경우에는 group by 적용 시 여러 개의 칼럼에 aggregation 함수를 호출하려면 대상 칼럼을 모두 Select 절에 나열하는 것이 다르다.
*   DataFrame의 groupby()에 특정 칼럼만 aggregation 함수를 적용하려면 groupby()로 반환된 DataFrameGroupBy 객체에 해당 칼럼을 필터링한 뒤 aggregation 함수를 적용한다.
*   항목 추가





In [75]:
titanic_df.count()

,0
PassengerId,891
Survived,891
Pclass,891
Name,891
Sex,891
Age,714
SibSp,891
Parch,891
Ticket,891
Fare,891


In [76]:
titanic_df[['Age', 'Fare']].mean()

,0
Age,29.699118
Fare,32.204208




*   DataFrame의 groupby()는 RDBMS SQL의 groupby 키워드와 유사하면서도 다른 면이 있기 때문에 주의가 필요하다. SQL과 판다스 모두 groupby를 분석 작업에 매우 많이 활용한다. DataFrame의 groupby() 사용 시 입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby된다. DataFrame에 groupby()를 호출하면 DataFrameGroupBy라는 또 다른 형태의 DataFrame을 반환한다.
*   SQL의 group by와 다르게, DataFrame에 groupby()를 호출해 반환한 결과에 aggregration 함수를 호출하면 groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용한다.



In [77]:
titanic_groupby = titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


In [78]:
titanic_groupby = titanic_df.groupby('Pclass').count()
titanic_groupby

,PassengerId,Survived,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
Pclass,,,,,,,,,,,
1,216,216,216,216,186,216,216,216,216,176,214
2,184,184,184,184,173,184,184,184,184,16,184
3,491,491,491,491,355,491,491,491,491,12,491


In [80]:
titanic_groupby = titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

,PassengerId,Survived
Pclass,,
1,216,216
2,184,184
3,491,491




*   또 다른 차이는 SQL의 경우 서로 다른 aggregation 함수를 적용할 경우에는 Select 절에 나열만 하면 되지만, DataFrame groupby()의 경우 적용하려는 여러 개의 aggregation 함수명을 DataFrameGroupBy 객체의 agg() 내에 인자로 입력해서 사용한다는 점이다.
*   그런데 이렇게 DataFrame의 groupby()를 이용해 API 기반으로 처리하다 보니 SQL의 group by보다 유연성이 떨어질 수 밖에 없다. gorupby()는 agg()를 이용해 이를 보완하는데, agg() 내에 입력값으로 딕셔너리 형태로 aggregation이 적용될 칼럼들과 aggregation 함수를 입력하면 된다.



In [81]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

/tmp/ipykernel_1000/1565262784.py:1: FutureWarning: The provided callable <built-in function max> is currently using SeriesGroupBy.max. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "max" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])
/tmp/ipykernel_1000/1565262784.py:1: FutureWarning: The provided callable <built-in function min> is currently using SeriesGroupBy.min. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "min" instead.
  titanic_df.groupby('Pclass')['Age'].agg([max, min])


,max,min
Pclass,,
1,80.0,0.92
2,70.0,0.67
3,74.0,0.42


In [82]:
agg_format = {'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

,Age,SibSp,Fare
Pclass,,,
1,80.0,90,84.154687
2,70.0,74,20.662183
3,74.0,302,13.675550


### 결손 데이터 처리하기



*   결손 데이터(Missing Data)는 칼럼에 값이 없는, 즉 NULL인 경우를 의미하며, 이를 넘파이의 NaN으로 표시한다. 기본적으로 머신러닝 알고리즘은 이 Nan 값을 처리하지 않으므로, 이 값을 다른 값으로 대체해야 한다. 또한 NaN 값은 평균, 총합 등의 함수 연산 시 제외된다. 예를 들어 특정 칼럼의 100개 데이터 중 10개가 NaN 값일 경우, 이 칼럼의 평균 값은 90개 데이터에 대한 평균이다.
*   NaN 여부를 확인하는 API는 isna()이며, NaN 값을 다른 값으로 대체하는 API는 fillna()이다. DataFrame에 isna()를 수행하면 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려준다. 결손 데이터의 개수는 isna() 결과에 sum() 함수를 추가해 구할 수 있다. sum()을 호출 시 True는 내부적으로 숫자 1로, False는 숫자 0으로 변환되므로 결손 데이터의 개수를 구할 수 있다.



In [83]:
titanic_df.isna().head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,False,False,False,False,False,False,False,False,False,False,True,False
1,False,False,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,True,False


In [84]:
titanic_df.isna().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,177
SibSp,0
Parch,0
Ticket,0
Fare,0




*   fillna()를 이용하면 결손 데이터를 편리하게 다른 값으로 대체할 수 있다. 주의해야 할 점은 filna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트 값이 변경된다는 점이다.



In [85]:
titanic_df['Cabin'] = titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr....",male,22.0,1,0,A/5 21171,7.2500,C000,S
1,2,1,1,"Cumings, Mr...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, ...",female,26.0,0,0,STON/O2. 31...,7.9250,C000,S


In [86]:
titanic_df['Age'] = titanic_df['Age'].fillna(titanic_df['Age'].mean()) # 'Age' 칼럼의 NaN 값을 평군 나이로 대체한다
titanic_df['Embarked'] = titanic_df['Embarked'].fillna('S') # 'Embarked' 칼럼의 NaN 값을 'S'로 대체한다
titanic_df.isna().sum()

,0
PassengerId,0
Survived,0
Pclass,0
Name,0
Sex,0
Age,0
SibSp,0
Parch,0
Ticket,0
Fare,0


### apply lambda 식으로 데이터 가공



*   판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제공한다. 판다스의 경우 칼럼에 일괄적으로 데이터 가공을 하는 것이 속도 면에서 더 빠르나, 복잡한 데이터 가공이 필요할 경우 어쩔 수 없이 apply lambda를 이용한다. lambda 식은 파이썬에서 함수형 프로그래밍(functional programming)을 지원하기 위해 만들어졌다.
*   사용자가 임의로 만드는 함수에서는 함수명과 입력 인자를 먼저 선언하고 이후 함수 내에서 입력 인자를 가공한 뒤 결괏값을 return과 같은 문법으로 반환해야 한다.lambda는 이러한 함수의 선언과 함수 내의 처리를 한 줄의 식으로 쉽게 변환하는 식이다. ':'로 입력 인자와 반환될 입력 인자의 계산식을 분리한다. ':'의 왼쪽에 있는 것은 입력 인자를 가리키며, 오른쪽은 입력 인자의 계산식, 즉 반환값을 의미한다. lambda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우에는 보통 map() 함수를 결합해 사용한다.



In [87]:
def get_square(a): # 입력값의 제곱 값을 구해서 반환하는 get_square(a)라는 함수 가정
  return a**2

print('3의 제곱은:', get_square(3))

3의 제곱은: 9


In [88]:
lambda_square = lambda x : x ** 2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


In [89]:
a = [1, 2, 3]
squares = map(lambda x : x**2, a)
list(squares)

[1, 4, 9]



*   판다스 DataFrame의 lambda 식은 파이썬의 이러한 lambda 식을 그대로 적용한 것이다. DataFrame의 apply에 lambda 식을 적용해 데이터를 가공하면 다음과 같다.
*   lambda식은 if else를 지원하는데, 주의할 점이 있다. if 절의 경우, if 식보다 반환 값을 먼저 기술해야 한다. 이는 lambda 식 ':' 기호 오른편에 반환 값이 있어야 하기 때문이다. 또 한 가지는, if와 else만 지원하고 if, else if, else와 같이 else if는 지원하지 않는다. else if를 이용하려면 else 절을 ()로 내포해 () 내에서 다시 if else 적용해 사용한다. 그런데 else if가 많이 나와야 하는 경우나 switch case 문의 경우에는 이렇게 else를 계속 내포해서 쓰기가 부담스럽다. 이 경우에는 아예 별도의 함수를 만드는 것이 더 나을 수 있다.



In [90]:
titanic_df['Name_len']= titanic_df['Name'].apply(lambda x : len(x)) # 'Name' 칼럼의 문자열 개수를 별도의 칼럼인 'Name_len'에 생성한다
titanic_df[['Name', 'Name_len']].head(3)

,Name,Name_len
0,"Braund, Mr....",23
1,"Cumings, Mr...",51
2,"Heikkinen, ...",22


In [92]:
titanic_df['Child_Adult'] = titanic_df['Age'].apply(lambda x : 'Child' if x <= 15 else 'Adult') # if else 절을 사용해 새로운 칼럼 'Child_Adult'를 생성한다
titanic_df[['Age', 'Child_Adult']].head(8)

,Age,Child_Adult
0,22.000000,Adult
1,38.000000,Adult
2,26.000000,Adult
3,35.000000,Adult
4,35.000000,Adult
5,29.699118,Adult
6,54.000000,Adult
7,2.000000,Child


In [93]:
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : 'Child' if x<=15 else ('Adult' if x <=60 else 'Elderly'))
titanic_df['Age_cat'].value_counts()

,count
Age_cat,
Adult,786
Child,83
Elderly,22


In [94]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성
def get_category(age):
  cat = ''
  if age <= 5: cat = 'Baby'
  elif age <= 12: cat = 'Child'
  elif age <= 18: cat = 'Teenager'
  elif age <= 25: cat = 'Student'
  elif age <= 35: cat = 'Young Adult'
  elif age <= 60: cat = 'Adult'
  else : cat = 'Elderly'

  return cat

# lambda 식에 위에서 생성한 get_category() 함수를 반환값으로 지정
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat'] = titanic_df['Age'].apply(lambda x : get_category(x))
titanic_df[['Age', 'Age_cat']].head()

,Age,Age_cat
0,22.0,Student
1,38.0,Adult
2,26.0,Young Adult
3,35.0,Young Adult
4,35.0,Young Adult
